File: server/test/local-api.test.js
'use strict';

/*
 * The INBOUND local REST door — the enablement half (Goal B part 3).
 *
 * ⚠️ WHY THIS FILE EXISTS AT ALL. The trigger feature shipped its definitions and its player code
 * complete and INERT: nothing wrote trigger_secret or the accept flags, so the secret was always
 * NULL, the resolver answered bad_secret to everything, and no listener ever bound. It looked
 * finished. A QA pass found a system that could not be switched on. These tests assert the switch.
 *
 * The assertions that matter here are the ones about what CANNOT happen:
 *   - the secret must never reach an API token, and never appear in a device list
 *   - enabling without a secret must be refused at the door, not left for the panel to fail quietly
 *   - the flag and the secret must be SEPARATE from the trigger ones, sharing a socket but not a
 *     permission — a site that wanted an emergency overlay must not thereby get remote control
 *   - the flag must ride every payload, so DISABLING takes effect
 */
const { test, before, after } = require('node:test');
const assert = require('node:assert/strict');
const { spawn } = require('node:child_process');
const path = require('node:path');
const os = require('node:os');
const fs = require('node:fs');
const crypto = require('node:crypto');
const ioClient = require('socket.io-client');

const { freePort } = require('./helpers/free-port');
const DATA_DIR = path.join(os.tmpdir(), 'st-lapi-' + crypto.randomBytes(4).toString('hex'));
const LOG = DATA_DIR + '.log';
let PORT, BASE, proc, jwt, workspaceId, deviceId, dbFile;

const sleep = (ms) => new Promise(r => setTimeout(r, ms));
const J = (tok, body, method = 'POST') => ({
  method, headers: { 'Content-Type': 'application/json', ...(tok ? { Authorization: `Bearer ${tok}` } : {}) },
  body: body === undefined ? undefined : JSON.stringify(body),
});
const H = (tok = jwt) => ({ headers: { Authorization: `Bearer ${tok}` } });
const dev = (p) => `${BASE}/api/devices/${deviceId}${p}`;

before(async () => {
  PORT = await freePort();
  BASE = `http://127.0.0.1:${PORT}`;
  const logFd = fs.openSync(LOG, 'w');
  proc = spawn('node', ['server.js'], {
    cwd: path.join(__dirname, '..'),
    env: { ...process.env, DATA_DIR, SELF_HOSTED: 'true', PORT: String(PORT), NODE_ENV: 'test' },
    stdio: ['ignore', logFd, logFd],
  });
  let up = false;
  for (let i = 0; i < 80; i++) {
    try { const r = await fetch(BASE + '/api/status'); if (r.ok) { up = true; break; } } catch { /* */ }
    await sleep(250);
  }
  if (!up) throw new Error('server did not boot:\n' + fs.readFileSync(LOG, 'utf8').slice(-2000));

  const reg = await (await fetch(BASE + '/api/auth/register', J(null, {
    email: `lapi${Date.now()}@example.com`, password: 'Passw0rd123', name: 'LAPI',
  }))).json();
  jwt = reg.token;
  workspaceId = reg.current_workspace_id;

  const Database = require('better-sqlite3');
  dbFile = path.join(DATA_DIR, 'db', 'remote_display.db');
  const raw = new Database(dbFile);
  deviceId = crypto.randomUUID();
  raw.prepare('INSERT INTO devices (id, name, workspace_id, status) VALUES (?, ?, ?, ?)')
    .run(deviceId, 'Lobby 1', workspaceId, 'offline');
  raw.close();
});
after(() => { try { proc.kill('SIGKILL'); } catch { /* */ } });

const row = () => {
  const Database = require('better-sqlite3');
  const raw = new Database(dbFile, { readonly: true });
  try { return raw.prepare('SELECT * FROM devices WHERE id = ?').get(deviceId); }
  finally { raw.close(); }
};

/* ------------------------------------------------------------------ the switch */

test('⚠️ enabling without a secret is refused AT THE DOOR', async () => {
  const r = await fetch(dev('/local-api'), J(jwt, { enabled: true }));
  assert.equal(r.status, 400);
  const b = await r.json();
  // The message has to name the fix. The panel would refuse every request with
  // 503 no_secret_configured, but an operator who ticked a box and walked away believes it is on.
  assert.match(b.error, /local-api-secret/);
  assert.equal(row().local_api_enabled, 0);
});

test('a secret can be set, and then the door can be opened', async () => {
  const s = await (await fetch(dev('/local-api-secret'), J(jwt, { rotate: true }))).json();
  assert.equal(s.success, true);
  assert.match(s.secret, /^[0-9a-f]{48}$/);          // 24 bytes hex — more than the trigger secret

  const e = await (await fetch(dev('/local-api'), J(jwt, { enabled: true }))).json();
  assert.equal(e.success, true);
  assert.equal(e.local_api.enabled, true);
  assert.equal(e.local_api.secret_set, true);
  assert.equal(row().local_api_enabled, 1);

  // The response names the command set, so an integrator is not guessing what the door accepts.
  assert.deepEqual(e.local_api.commands.sort(),
    ['refresh', 'screen_off', 'screen_on', 'set_brightness', 'set_system_brightness', 'set_volume']);
  // ...and the port, because it is the TRIGGER port. Two doors, one socket.
  assert.equal(e.local_api.port, 8079);
});

test('a short secret is refused; the floor is offline-guessability, not policy', async () => {
  const r = await fetch(dev('/local-api-secret'), J(jwt, { secret: 'short' }));
  assert.equal(r.status, 400);
  const r2 = await fetch(dev('/local-api-secret'), J(jwt, { secret: 'has a space in it and is long enough' }));
  assert.equal(r2.status, 400);
});

test('it can be turned back OFF, which is the half that is usually missing', async () => {
  await fetch(dev('/local-api'), J(jwt, { enabled: true }));
  const e = await (await fetch(dev('/local-api'), J(jwt, { enabled: false }))).json();
  assert.equal(e.local_api.enabled, false);
  assert.equal(row().local_api_enabled, 0);
  // Re-enable for the payload test below; the secret is still set, so this must succeed.
  await fetch(dev('/local-api'), J(jwt, { enabled: true }));
});

/* ------------------------------------------------------------------ secrecy */

test('⚠️ the secret never appears in the device list', async () => {
  const list = await (await fetch(BASE + '/api/devices', H())).json();
  const blob = JSON.stringify(list);
  assert.ok(!blob.includes(row().local_api_secret), 'the local API secret escaped into the list');
  const arr = Array.isArray(list) ? list : list.devices;
  assert.equal(arr.find((d) => d.id === deviceId).local_api_secret, undefined);
});

test('⚠️ the secret never reaches an API token, even a full-scope one', async () => {
  const t = await (await fetch(BASE + '/api/tokens', J(jwt, { name: 'ci', scope: 'full' }))).json();
  const token = t.token || (t.api_token && t.api_token.token);
  assert.ok(token, 'no token issued: ' + JSON.stringify(t));

  // Detail: the endpoint with no scope gate, which is where the enrol-key version of this bug lived.
  const d = await (await fetch(`${BASE}/api/devices/${deviceId}`, H(token))).json();
  const device = d.device || d;
  assert.equal(device.local_api_secret, undefined);
  assert.equal(device.trigger_secret, undefined);

  // And rotation through a token says so rather than handing the value over.
  const rot = await (await fetch(dev('/local-api-secret'), J(token, { rotate: true }))).json();
  assert.equal(rot.secret, undefined);
  assert.equal(rot.secret_set, true);
  assert.match(rot.note, /not returned to API tokens/);
});

/* ------------------------------------------------------- separate from triggers */

test('⚠️ the trigger flag and the local-api flag are INDEPENDENT', async () => {
  // Triggers off, control API on: the site that wants remote control but no overlays.
  await fetch(dev('/trigger-config'), J(jwt, { accept_http: false, accept_udp: false }));
  await fetch(dev('/local-api'), J(jwt, { enabled: true }));
  let r = row();
  assert.equal(r.triggers_accept_http, 0);
  assert.equal(r.local_api_enabled, 1);

  // And the reverse: the emergency-overlay site that must NOT get remote control thrown in.
  await fetch(dev('/local-api'), J(jwt, { enabled: false }));
  await fetch(dev('/trigger-config'), J(jwt, { accept_http: true }));
  r = row();
  assert.equal(r.triggers_accept_http, 1);
  assert.equal(r.local_api_enabled, 0, 'enabling triggers must not enable the control door');
});

test('⚠️ the secrets are separate values, so revoking one does not revoke the other', async () => {
  await fetch(dev('/trigger-secret'), J(jwt, { rotate: true }));
  const before = row();
  assert.ok(before.trigger_secret && before.local_api_secret);
  assert.notEqual(before.trigger_secret, before.local_api_secret);

  await fetch(dev('/local-api-secret'), J(jwt, { rotate: true }));
  const after = row();
  assert.equal(after.trigger_secret, before.trigger_secret, 'rotating one rotated the other');
  assert.notEqual(after.local_api_secret, before.local_api_secret);
});

/* ------------------------------------------------------------------ the payload */

/*
 * ⚠️ ASSERTED ON WHAT A REAL DEVICE RECEIVES, not on an in-process buildPlaylistPayload().
 *
 * That helper is only exported once setupDeviceSocket(io) has run, so requiring it here would read
 * a different database from the one the routes just wrote — which is exactly the mistake that made
 * two earlier tests in this feature assert on an empty DB and pass for the wrong reason. A socket
 * client costs a few lines and tests the thing the panel actually gets.
 */
function registerDevice() {
  const code = String(crypto.randomInt(100000, 1000000));
  return new Promise((resolve) => {
    const sock = ioClient(`${BASE}/device`, { transports: ['websocket'], reconnection: false, forceNew: true });
    sock.on('connect', () => sock.emit('device:register', { pairing_code: code }));
    sock.on('device:registered', (d) => resolve({ sock, id: d.device_id, token: d.device_token }));
    setTimeout(() => resolve(null), 5000);
  });
}

/** A self-provisioned device has no workspace; the routes under test are workspace-scoped. */
function claim(id) {
  const Database = require('better-sqlite3');
  const raw = new Database(dbFile);
  try { raw.prepare('UPDATE devices SET workspace_id = ? WHERE id = ?').run(workspaceId, id); }
  finally { raw.close(); }
}

/**
 * Wait for a playlist-update that SATISFIES the predicate, not merely the next one.
 *
 * ⚠️ `once` was wrong and failed for the right reason: a device that has just registered already has
 * a payload in flight, so the first update to arrive after a toggle is frequently the one from
 * before it. A test that asserts on "the next payload" is a test that passes or fails on timing.
 */
function payloadWhere(sock, pred, ms = 5000) {
  return new Promise((resolve) => {
    const seen = [];
    const onUpdate = (p) => {
      seen.push(p && p.local_api);
      if (pred(p)) { cleanup(); resolve({ payload: p, seen }); }
    };
    const cleanup = () => { clearTimeout(t); sock.off('device:playlist-update', onUpdate); };
    const t = setTimeout(() => { cleanup(); resolve({ payload: null, seen }); }, ms);
    sock.on('device:playlist-update', onUpdate);
  });
}

test('⚠️ the payload carries local_api — and carries it when OFF, too', async () => {
  const d = await registerDevice();
  assert.ok(d, 'the device never registered');
  claim(d.id);
  try {
    const target = (p) => `${BASE}/api/devices/${d.id}${p}`;
    const sr = await fetch(target('/local-api-secret'), J(jwt, { rotate: true }));
    assert.equal(sr.status, 200, 'secret rotation on the socket device: ' + (await sr.text()).slice(0, 200));

    const onWait = payloadWhere(d.sock, (p) => p && p.local_api && p.local_api.enabled === true);
    const er = await fetch(target('/local-api'), J(jwt, { enabled: true }));
    const eb = await er.json();
    assert.equal(er.status, 200, 'enable: ' + JSON.stringify(eb));
    assert.equal(eb.local_api.enabled, true, 'route echo: ' + JSON.stringify(eb));
    assert.equal(eb.delivered, true, 'the push was queued, not emitted: ' + JSON.stringify(eb));
    const onResult = await onWait;
    const on = onResult.payload;
    assert.ok(on, 'no payload with local_api.enabled arrived; saw ' + JSON.stringify(onResult.seen));
    // The panel cannot compare a credential it was never given.
    assert.ok(on.local_api.secret && on.local_api.secret.length >= 16);

    const offWait = payloadWhere(d.sock, (p) => p && p.local_api && p.local_api.enabled === false);
    await fetch(target('/local-api'), J(jwt, { enabled: false }));
    const offResult = await offWait;
    const off = offResult.payload;
    assert.ok(off, 'no payload with the door closed arrived: ' + JSON.stringify(offResult.seen));
    /*
     * ⚠️ The field must still be PRESENT when disabled. A panel that only ever learned "on" would
     * keep the door open for the rest of its life — absence has to mean off, and the only way to
     * make absence mean off is to always send the field.
     */
    assert.ok(off.local_api, 'local_api vanished when disabled, so a panel could never be told to close');
    assert.equal(off.local_api.enabled, false);

    /*
     * ⚠️ TOP-LEVEL, outside the item list. Same rule as triggers, the power schedule and saved
     * endpoints: toggling a door must not read as a content change and restart playback (#234).
     */
    assert.ok(Object.prototype.hasOwnProperty.call(off, 'local_api'));
    assert.ok(!JSON.stringify(off.assignments || []).includes('local_api'));
  } finally {
    try { d.sock.close(); } catch { /* */ }
  }
});

File: android/app/src/test/java/com/remotedisplay/player/net/LocalApiTest.kt
package com.remotedisplay.player.net

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The inbound local REST door's DECISION half (Goal B part 3).
 *
 * ⚠️ These are the assertions that cannot be made once the logic lives in a socket handler, which is
 * the whole reason LocalApi is separable from TriggerListeners: a refusal is only testable if you can
 * ask for one without binding a port.
 *
 * The cases worth having are all about what must NOT happen:
 *   - the door must be invisible when disabled, not merely closed
 *   - a panel with the flag on and no secret must refuse everything, not accept anything
 *   - a command the panel implements but that is not on the list must be refused, and refused with a
 *     status that sends an integrator to their permissions rather than their spelling
 *   - the status body must not be reachable without the secret, because it is the cheapest thing on
 *     the door and therefore the first thing anyone tries
 */
class LocalApiTest {

    private val cfg = LocalApi.Config(enabled = true, secret = "0123456789abcdef0123")
    private val status = { JSONObject().put("ok", true).put("name", "Lobby 1") }

    private fun get(path: String, q: Map<String, String> = emptyMap(), h: Map<String, String> = emptyMap(),
                    c: LocalApi.Config = cfg) =
        LocalApi.handle("GET", path, q, h, "", c, status)

    private fun post(path: String, body: String, q: Map<String, String> = emptyMap(),
                     h: Map<String, String> = mapOf("authorization" to "Bearer 0123456789abcdef0123"),
                     c: LocalApi.Config = cfg) =
        LocalApi.handle("POST", path, q, h, body, c, status)

    private fun auth() = mapOf("authorization" to "Bearer 0123456789abcdef0123")

    /* ------------------------------------------------------------------ the door */

    @Test
    fun `disabled is INVISIBLE, not forbidden`() {
        // ⚠️ 404 and not 403. A 403 advertises the feature to anyone scanning the segment: it says
        // "this panel has a control API and you have the wrong key". A panel that was never switched
        // on must look like a panel that has no such door.
        val r = get("/api/status", h = auth(), c = LocalApi.Config(enabled = false, secret = "x".repeat(20)))
        assertEquals(404, r.status)
        assertTrue(r.body.contains("not_found"))
        assertNull(r.command)
    }

    @Test
    fun `⚠️ enabled with NO secret refuses everything`() {
        /*
         * The trap TriggerManager documents for the string "null": a panel switched on before it was
         * configured must be CLOSED for the window in between, not open. A fleet enabled by a script
         * and secured later is exactly how that window happens.
         */
        val r = get("/api/status", h = auth(), c = LocalApi.Config(enabled = true, secret = null))
        assertEquals(503, r.status)
        assertTrue(r.body.contains("no_secret_configured"))
        val r2 = get("/api/status", h = auth(), c = LocalApi.Config(enabled = true, secret = ""))
        assertEquals(503, r2.status)
    }

    /* ------------------------------------------------------------------ the secret */

    @Test
    fun `status is NOT readable without the secret`() {
        // The cheapest request on the door, so the first one anybody tries.
        assertEquals(401, get("/api/status").status)
        assertEquals(401, get("/api/status", h = mapOf("authorization" to "Bearer wrong")).status)
        val ok = get("/api/status", h = auth())
        assertEquals(200, ok.status)
        assertTrue(ok.body.contains("Lobby 1"))
    }

    @Test
    fun `the secret is accepted as a bearer header, a bare header, or a query parameter`() {
        assertEquals(200, get("/api/status", h = mapOf("authorization" to "Bearer 0123456789abcdef0123")).status)
        // Bare, because plenty of gear writes the header but not the scheme.
        assertEquals(200, get("/api/status", h = mapOf("authorization" to "0123456789abcdef0123")).status)
        // Case-insensitive scheme: hand-concatenated HTTP gets this wrong constantly.
        assertEquals(200, get("/api/status", h = mapOf("authorization" to "bearer 0123456789abcdef0123")).status)
        // ⚠️ Query, for AMX NetLinx and Extron Global Scripter, which cannot set a header at all.
        assertEquals(200, get("/api/status", q = mapOf("secret" to "0123456789abcdef0123")).status)
    }

    @Test
    fun `an EMPTY header falls through to the query rather than failing outright`() {
        // A sender that emits `Authorization:` with nothing after it would otherwise be refused even
        // though it also sent a perfectly good ?secret= — a confusing 401 on a correct request.
        val r = get("/api/status", q = mapOf("secret" to "0123456789abcdef0123"),
                    h = mapOf("authorization" to ""))
        assertEquals(200, r.status)
    }

    @Test
    fun `a secret of a different length is refused without comparing content`() {
        assertEquals(401, get("/api/status", h = mapOf("authorization" to "Bearer 0123456789abcdef")).status)
        assertEquals(401, get("/api/status", h = mapOf("authorization" to "Bearer 0123456789abcdef01234567")).status)
    }

    /* ------------------------------------------------------------------ commands */

    @Test
    fun `an allowed command is returned for the caller to run, and answered 200`() {
        val r = post("/api/command", """{"type":"screen_off"}""")
        assertEquals(200, r.status)
        assertEquals("screen_off", r.command)
        assertTrue(r.body.contains("\"ok\":true"))
    }

    @Test
    fun `⚠️ a real command that is not on the list is 403, not 400`() {
        /*
         * `reboot` IS a command this panel implements. Answering "invalid command" would send an
         * integrator to check their spelling, when the honest answer is that this door does not grant
         * it. 400 and 403 send someone to two different places, and only one of them is the right one.
         */
        for (t in listOf("reboot", "shell", "install_apk", "set_server_url", "http_request", "launch",
                         "kiosk_unlock", "settings", "set_power_schedule", "update")) {
            val r = post("/api/command", """{"type":"$t"}""")
            assertEquals("$t must be refused", 403, r.status)
            assertNull("$t must not be handed to the dispatcher", r.command)
            assertTrue(r.body.contains("command_not_permitted"))
        }
    }

    @Test
    fun `the payload is taken nested OR flat`() {
        // ⚠️ Both, because a control system that hand-concatenates JSON gets one level of nesting
        // right and two wrong — and the nested form failing SILENTLY is the worst outcome: a valid
        // set_volume with an empty payload is a volume of zero.
        val nested = post("/api/command", """{"type":"set_volume","payload":{"volume":40}}""")
        assertEquals(40, nested.payload?.optInt("volume"))
        val flat = post("/api/command", """{"type":"set_volume","volume":40}""")
        assertEquals(40, flat.payload?.optInt("volume"))
        // And `type` is not left in the flat payload, where a handler might read it as data.
        assertFalse(flat.payload!!.has("type"))
    }

    @Test
    fun `a non-JSON body can still name a command on the query`() {
        // Same gear the trigger door's GET support exists for.
        val r = LocalApi.handle("POST", "/api/command", mapOf("type" to "refresh"), auth(), "not json at all", cfg, status)
        assertEquals(200, r.status)
        assertEquals("refresh", r.command)
    }

    @Test
    fun `a command with no type is 400`() {
        assertEquals(400, post("/api/command", """{}""").status)
        assertEquals(400, post("/api/command", """{"type":""}""").status)
    }

    @Test
    fun `⚠️ a wrong secret on a command never reaches the command`() {
        val r = LocalApi.handle("POST", "/api/command", emptyMap(), mapOf("authorization" to "Bearer nope"),
                                """{"type":"screen_off"}""", cfg, status)
        assertEquals(401, r.status)
        assertNull(r.command)
    }

    /* ------------------------------------------------------------------ methods and paths */

    @Test
    fun `the methods are fixed per path`() {
        assertEquals(405, LocalApi.handle("POST", "/api/status", emptyMap(), auth(), "", cfg, status).status)
        assertEquals(405, LocalApi.handle("GET", "/api/command", emptyMap(), auth(), "", cfg, status).status)
    }

    @Test
    fun `only the two paths are this door's`() {
        assertTrue(LocalApi.isLocalApiPath("/api/status"))
        assertTrue(LocalApi.isLocalApiPath("/api/command"))
        // ⚠️ Everything else on that socket is a TRIGGER. A near-miss must fall through rather than
        // be answered here, or a site POSTing triggers to /api/commands stops working.
        for (p in listOf("/", "/api", "/api/commands", "/api/status/", "/API/status", "/trigger")) {
            assertFalse(p, LocalApi.isLocalApiPath(p))
        }
    }

    /* ------------------------------------------------------------------ the status body */

    @Test
    fun `a status builder that throws yields an empty body, not a 500`() {
        // A control system polling status must not be able to take the door down with a panel-side
        // bug, and "{}" is a far better answer than a dropped connection.
        val r = LocalApi.handle("GET", "/api/status", emptyMap(), auth(), "", cfg) { throw RuntimeException("boom") }
        assertEquals(200, r.status)
        assertEquals("{}", r.body)
    }

    @Test
    fun `the command list is the room-control set and nothing more`() {
        assertEquals(
            listOf("refresh", "screen_on", "screen_off", "set_volume", "set_brightness", "set_system_brightness").sorted(),
            LocalApi.COMMANDS.sorted()
        )
    }
}
